### ISSUE TITLE
Add a named pipe support to `Server`

### ISSUE THREAD

**Issue by karthiknadig** [openlawlibrary/pygls#403]
The IO mode is sometimes impacted be certain libraries writing directly to stdout. We have seen this occur with python logging library that has been configured to output to stdout. 

With sockets, users have firewall rules that can sometimes prevent sockets from being opened.

The language client for VS Code supports using named pipe, where a named pipe is created on the node side and passes it into the server.

The feature request is to add a `start_pipe` method to `Server`

**Comment by alcarney** [comment]
I'm not familiar with named pipes... how would they look in Python? Would the server basically be given a pair of "files" (`/dev/in_pipe`, `/dev/out_pipe`) that it could read/write to?

Is this something we can support with the existing [`start_io`](https://github.com/openlawlibrary/pygls/blob/8e1e8fa3b1ab16fcb804f7d5330ece2e5583206b/pygls/server.py#L247) function by passing in the relevant objects? e.g. `server.start_io(stdin=in_pipe, stdout=out_pipe)`

**Comment by karthiknadig** [comment]
You can treat the name pipe like a file in python. So you can use `open` in python to use it. I will add more technical details on this, as there may be subtle differences between windows vs linux on how the pipes are handled.

**Comment by karthiknadig** [comment]
This seems to be working for me:
```py
import argparse
import os
from pygls import server

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('--pipe', action='store', dest='pipe', help='Named pipe to connect to', required=True)
    return parser.parse_args()

LSP_SERVER = server.LanguageServer(
    name="demo-server", version="v0.1.0", max_workers=5
)

if __name__ == "__main__":
    named_pipe = parse_args().pipe
    fd = os.open(named_pipe, os.O_RDWR)
    with os.fdopen(os.dup(fd), 'rb') as r:
        with os.fdopen(os.dup(fd), 'wb') as w:
            LSP_SERVER.start_io(r, w)
```

I have only tested this on windows.

```console
2023-10-24 19:27:08.694 [info] [Trace - 7:27:08 PM] Sending request 'initialize - (0)'.
2023-10-24 19:27:08.695 [info] Params: {
    "processId": 23664,
    "clientInfo": {
        "name": "Visual Studio Code - Insiders",
        "version": "1.84.0-insider"
    },
    "locale": "en",
    "rootPath": "c:\\GIT\\demo\\diff_test",
    "rootUri": "file:///c%3A/GIT/demo/diff_test",
    "capabilities": {
        "workspace": {
            "applyEdit": true,
            "workspaceEdit": {
                "documentChanges": true,
                "resourceOperations": [
                    "create",
                    "rename",
                    "delete"
                ],
                "failureHandling": "textOnlyTransactional",
                "normalizesLineEndings": true,
                "changeAnnotationSupport": {
                    "groupsOnLabel": true
                }
            },
            "configuration": true,
            "didChangeWatchedFiles": {
                "dynamicRegistration": true,
                "relativePatternSupport": true
            },
            "symbol": {
                "dynamicRegistration": true,
                "symbolKind": {
                    "valueSet": [
                        1,
                        2,
                        3,
                        4,
                        5,
                        6,
                        7,
                        8,
                        9,
                        10,
                        11,
                        12,
                        13,
                        14,
                        15,
                        16,
                        17,
                        18,
                        19,
                        20,
                        21,
                        22,
                        23,
                        24,
                        25,
                        26
                    ]
                },
                "tagSupport": {
                    "valueSet": [
                        1
                    ]
                },
                "resolveSupport": {
                    "properties": [
                        "location.range"
                    ]
                }
            },
            "codeLens": {
                "refreshSupport": true
            },
            "executeCommand": {
                "dynamicRegistration": true
            },
            "didChangeConfiguration": {
                "dynamicRegistration": true
            },
            "workspaceFolders": true,
            "semanticTokens": {
                "refreshSupport": true
            },
            "fileOperations": {
                "dynamicRegistration": true,
                "didCreate": true,
                "didRename": true,
                "didDelete": true,
                "willCreate": true,
                "willRename": true,
                "willDelete": true
            },
            "inlineValue": {
                "refreshSupport": true
            },
            "inlayHint": {
                "refreshSupport": true
            },
            "diagnostics": {
                "refreshSupport": true
            }
        },
        "textDocument": {
            "publishDiagnostics": {
                "relatedInformation": true,
                "versionSupport": false,
                "tagSupport": {
                    "valueSet": [
                        1,
                        2
                    ]
                },
                "codeDescriptionSupport": true,
                "dataSupport": true
            },
            "synchronization": {
                "dynamicRegistration": true,
                "willSave": true,
                "willSaveWaitUntil": true,
                "didSave": true
            },
            "completion": {
                "dynamicRegistration": true,
                "contextSupport": true,
                "completionItem": {
                    "snippetSupport": true,
                    "commitCharactersSupport": true,
                    "documentationFormat": [
                        "markdown",
                        "plaintext"
                    ],
                    "deprecatedSupport": true,
                    "preselectSupport": true,
                    "tagSupport": {
                        "valueSet": [
                            1
                        ]
                    },
                    "insertReplaceSupport": true,
                    "resolveSupport": {
                        "properties": [
                            "documentation",
                            "detail",
                            "additionalTextEdits"
                        ]
                    },
                    "insertTextModeSupport": {
                        "valueSet": [
                            1,
                            2
                        ]
                    },
                    "labelDetailsSupport": true
                },
                "insertTextMode": 2,
                "completionItemKind": {
                    "valueSet": [
                        1,
                        2,
                        3,
                        4,
                        5,
                        6,
                        7,
                        8,
                        9,
                        10,
                        11,
                        12,
                        13,
                        14,
                        15,
                        16,
                        17,
                        18,
                        19,
                        20,
                        21,
                        22,
                        23,
                        24,
                        25
                    ]
                },
                "completionList": {
                    "itemDefaults": [
                        "commitCharacters",
                        "editRange",
                        "insertTextFormat",
                        "insertTextMode"
                    ]
                }
            },
            "hover": {
                "dynamicRegistration": true,
                "contentFormat": [
                    "markdown",
                    "plaintext"
                ]
            },
            "signatureHelp": {
                "dynamicRegistration": true,
                "signatureInformation": {
                    "documentationFormat": [
                        "markdown",
                        "plaintext"
                    ],
                    "parameterInformation": {
                        "labelOffsetSupport": true
                    },
                    "activeParameterSupport": true
                },
                "contextSupport": true
            },
            "definition": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "references": {
                "dynamicRegistration": true
            },
            "documentHighlight": {
                "dynamicRegistration": true
            },
            "documentSymbol": {
                "dynamicRegistration": true,
                "symbolKind": {
                    "valueSet": [
                        1,
                        2,
                        3,
                        4,
                        5,
                        6,
                        7,
                        8,
                        9,
                        10,
                        11,
                        12,
                        13,
                        14,
                        15,
                        16,
                        17,
                        18,
                        19,
                        20,
                        21,
                        22,
                        23,
                        24,
                        25,
                        26
                    ]
                },
                "hierarchicalDocumentSymbolSupport": true,
                "tagSupport": {
                    "valueSet": [
                        1
                    ]
                },
                "labelSupport": true
            },
            "codeAction": {
                "dynamicRegistration": true,
                "isPreferredSupport": true,
                "disabledSupport": true,
                "dataSupport": true,
                "resolveSupport": {
                    "properties": [
                        "edit"
                    ]
                },
                "codeActionLiteralSupport": {
                    "codeActionKind": {
                        "valueSet": [
                            "",
                            "quickfix",
                            "refactor",
                            "refactor.extract",
                            "refactor.inline",
                            "refactor.rewrite",
                            "source",
                            "source.organizeImports"
                        ]
                    }
                },
                "honorsChangeAnnotations": false
            },
            "codeLens": {
                "dynamicRegistration": true
            },
            "formatting": {
                "dynamicRegistration": true
            },
            "rangeFormatting": {
                "dynamicRegistration": true
            },
            "onTypeFormatting": {
                "dynamicRegistration": true
            },
            "rename": {
                "dynamicRegistration": true,
                "prepareSupport": true,
                "prepareSupportDefaultBehavior": 1,
                "honorsChangeAnnotations": true
            },
            "documentLink": {
                "dynamicRegistration": true,
                "tooltipSupport": true
            },
            "typeDefinition": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "implementation": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "colorProvider": {
                "dynamicRegistration": true
            },
            "foldingRange": {
                "dynamicRegistration": true,
                "rangeLimit": 5000,
                "lineFoldingOnly": true,
                "foldingRangeKind": {
                    "valueSet": [
                        "comment",
                        "imports",
                        "region"
                    ]
                },
                "foldingRange": {
                    "collapsedText": false
                }
            },
            "declaration": {
                "dynamicRegistration": true,
                "linkSupport": true
            },
            "selectionRange": {
                "dynamicRegistration": true
            },
            "callHierarchy": {
                "dynamicRegistration": true
            },
            "semanticTokens": {
                "dynamicRegistration": true,
                "tokenTypes": [
                    "namespace",
                    "type",
                    "class",
                    "enum",
                    "interface",
                    "struct",
                    "typeParameter",
                    "parameter",
                    "variable",
                    "property",
                    "enumMember",
                    "event",
                    "function",
                    "method",
                    "macro",
                    "keyword",
                    "modifier",
                    "comment",
                    "string",
                    "number",
                    "regexp",
                    "operator",
                    "decorator"
                ],
                "tokenModifiers": [
                    "declaration",
                    "definition",
                    "readonly",
                    "static",
                    "deprecated",
                    "abstract",
                    "async",
                    "modification",
                    "documentation",
                    "defaultLibrary"
                ],
                "formats": [
                    "relative"
                ],
                "requests": {
                    "range": true,
                    "full": {
                        "delta": true
                    }
                },
                "multilineTokenSupport": false,
                "overlappingTokenSupport": false,
                "serverCancelSupport": true,
                "augmentsSyntaxTokens": true
            },
            "linkedEditingRange": {
                "dynamicRegistration": true
            },
            "typeHierarchy": {
                "dynamicRegistration": true
            },
            "inlineValue": {
                "dynamicRegistration": true
            },
            "inlayHint": {
                "dynamicRegistration": true,
                "resolveSupport": {
                    "properties": [
                        "tooltip",
                        "textEdits",
                        "label.tooltip",
                        "label.location",
                        "label.command"
                    ]
                }
            },
            "diagnostic": {
                "dynamicRegistration": true,
                "relatedDocumentSupport": false
            }
        },
        "window": {
            "showMessage": {
                "messageActionItem": {
                    "additionalPropertiesSupport": true
                }
            },
            "showDocument": {
                "support": true
            },
            "workDoneProgress": true
        },
        "general": {
            "staleRequestSupport": {
                "cancel": true,
                "retryOnContentModified": [
                    "textDocument/semanticTokens/full",
                    "textDocument/semanticTokens/range",
                    "textDocument/semanticTokens/full/delta"
                ]
            },
            "regularExpressions": {
                "engine": "ECMAScript",
                "version": "ES2020"
            },
            "markdown": {
                "parser": "marked",
                "version": "1.1.0"
            },
            "positionEncodings": [
                "utf-16"
            ]
        },
        "notebookDocument": {
            "synchronization": {
                "dynamicRegistration": true,
                "executionSummarySupport": true
            }
        }
    },
    "trace": "verbose",
    "workspaceFolders": [
        {
            "uri": "file:///c%3A/GIT/demo/diff_test",
            "name": "diff_test"
        }
    ]
}


2023-10-24 19:27:10.855 [info] [Trace - 7:27:10 PM] Received response 'initialize - (0)' in 2332ms.
2023-10-24 19:27:10.855 [info] Result: {
    "capabilities": {
        "positionEncoding": "utf-16",
        "textDocumentSync": {
            "openClose": true,
            "change": 2,
            "willSave": false,
            "willSaveWaitUntil": false,
            "save": false
        },
        "executeCommandProvider": {
            "commands": []
        },
        "diagnosticProvider": {
            "interFileDependencies": false,
            "workspaceDiagnostics": false
        },
        "workspace": {
            "workspaceFolders": {
                "supported": true,
                "changeNotifications": true
            },
            "fileOperations": {}
        }
    },
    "serverInfo": {
        "name": "demo-server",
        "version": "v0.1.0"
    }
}


2023-10-24 19:27:10.855 [info] [Trace - 7:27:10 PM] Sending notification 'initialized'.
2023-10-24 19:27:10.855 [info] Params: {}


2023-10-24 19:27:12.981 [info] [Trace - 7:27:12 PM] Sending notification 'textDocument/didOpen'.
2023-10-24 19:27:12.981 [info] Params: {
    "textDocument": {
        "uri": "file:///c%3A/GIT/demo/diff_test/querry_count.py",
        "languageId": "python",
        "version": 1,
        "text": "import urllib.request as url_lib\r\nfrom bs4 import BeautifulSoup\r\nimport re\r\n\r\nMARKETPLACE_ROOT = \"https://marketplace.visualstudio.com/items?itemName=\"\r\nMARKETPLACE = [\r\n\"d-biehl.robotcode\",\r\n\"broadinstitute.wdl-devtools\",\r\n\"ThierryJean.hamilton\",\r\n\"oliversen.chatgpt-docstrings\",\r\n\"OpShinDev.opshin\",\r\n\"AntiMach.symdexer\",\r\n\"compilation.compilation\",\r\n\"olleln.pyrefact\",\r\n\"charliermarsh.ruff\",\r\n\"mikoz.autoflake-extension\",\r\n\"omnilib.ufmt\",\r\n\"Explosion.spacy-extension\",\r\n\"davelopez.galaxy-tools\",\r\n\"CrossHair.crosshair\",\r\n\"avast-threatlabs-yara.vscode-yls\",\r\n\"uniteai.uniteai\",\r\n\"h2oai.h2o-wave\",\r\n\"hai-bo.stata-language-server\",\r\n\"jaseci-labs.jac\",\r\n\"perrinjerome.vscode-zc-buildout\",\r\n\"biometria-se.grizzly-loadtester-vscode\",\r\n\"Siithub.spm\",\r\n\"dancermak.salt-lsp\",\r\n\"matangover.mypy\",\r\n]\r\n\r\n\r\nFIRST_PARTY = [\r\n\"ms-python.pylint\",\r\n\"ms-python.black-formatter\",\r\n\"ms-python.autopep8\",\r\n#\"ms-python.isort\",\r\n\"ms-python.flake8\",\r\n\"ms-python.mypy-type-checker\",\r\n]\r\n\r\n\r\nPYPI_ROOT = \"https://pypistats.org/packages/\"\r\nPYPI = [\r\n\"lsprotocol\",\r\n\"jedi-language-server\",\r\n\"ruff-lsp\",\r\n\"helios-language-server\",\r\n\"anakin-language-server\",\r\n\"pkgbuild-language-server\",\r\n\"esbonio\",\r\n\"textLSP\",\r\n\"cp2k-input-tools\",\r\n\"cfn-lsp-extra\",\r\n\"autotools-language-server\",\r\n\"nginx-language-server\",\r\n\"dmypy-ls\",\r\n\"pytest-yls\",\r\n\"yls\",\r\n]\r\n\r\nGH_ROOT = \"https://github.com/\"\r\nGH = [\r\n\"danixeee/textx-gen-vscode\",\r\n\"ompugao/markshift\",\r\n\"Mastermori/DemonCore\",\r\n\"swyddfa/lsp-devtools\",\r\n\"oskarissimus/antlr-playground\",\r\n\"AdaCore/RecordFlux\",\r\n\"msc94/bazel-language-server\",\r\n\"georgeneokq/uNGINXed-linter\",\r\n]\r\n\r\n# get download count of VS Code marketplace extension\r\ndef get_download_count(url:str) -> int :\r\n\r\n    with url_lib.urlopen(url) as response:\r\n        data = response.read()\r\n        soup = BeautifulSoup(data, \"html.parser\")\r\n        download_count = soup.find(\"span\", class_=\"installs-text\")\r\n        download_count = re.sub(r\"[^\\d]\", \"\", download_count.text)\r\n    return int(download_count)\r\n\r\n\r\n# get stats for the pipy package\r\ndef get_pypi_stats(url:str) -> int :\r\n    with url_lib.urlopen(url) as response:\r\n        data = response.read()\r\n        soup = BeautifulSoup(data, \"html.parser\")\r\n        section = soup.find(\"section\")\r\n        if section is None:\r\n            return 0\r\n        \r\n        p = section.find(\"p\")\r\n        if p is None:\r\n            return 0\r\n        # extract patter:\r\n        # Downloads last month: 17,974,805\r\n        pattern = r\"Downloads last month:\\s*(\\d*,?\\d*,?\\d+)\"\r\n        matches = re.findall(pattern, p.text)\r\n        if not matches:\r\n            return 0\r\n        download_count = matches[0].replace(\",\", \"\")\r\n    return int(download_count)\r\n\r\nprint(\"VS Code Extensions 1st party (unique installs):\")\r\ntotal = 0\r\nfor item in FIRST_PARTY:\r\n    count = get_download_count(MARKETPLACE_ROOT + item)\r\n    total += count\r\n    print(f\"{item}: {count}\")\r\nprint(\"--------------\")\r\nprint(f\"Total: {total}\")\r\nprint(\"--------------\")\r\n\r\nprint(\"VS Code Extensions 3rd party (unique installs):\")\r\ntotal = 0\r\nfor item in MARKETPLACE:\r\n    count = get_download_count(MARKETPLACE_ROOT + item)\r\n    total += count\r\n    print(f\"{item}: {count}\")\r\nprint(\"--------------\")\r\nprint(f\"Total: {total}\")\r\nprint(\"--------------\")\r\n\r\nprint(\"PyPi Packages (monthly downloads):\")\r\ntotal = 0\r\nfor item in PYPI:\r\n    count = get_pypi_stats(PYPI_ROOT + item)\r\n    total += count\r\n    print(f\"{item}: {count}\")\r\nprint(\"--------------\")\r\nprint(f\"Total: {total}\")\r\nprint(\"--------------\")\r\n"
    }
}


2023-10-24 19:27:12.981 [info] [Trace - 7:27:12 PM] Sending request 'textDocument/diagnostic - (1)'.
2023-10-24 19:27:12.981 [info] Params: {
    "textDocument": {
        "uri": "file:///c%3A/GIT/demo/diff_test/querry_count.py"
    }
}


2023-10-24 19:27:13.528 [info] [Trace - 7:27:13 PM] Received response 'textDocument/diagnostic - (1)' in 547ms.
2023-10-24 19:27:13.529 [info] Result: {
    "items": [
        {
            "range": {
                "start": {
                    "line": 0,
                    "character": 0
                },
                "end": {
                    "line": 0,
                    "character": 1
                }
            },
            "message": "This is a demo diagnostic",
            "severity": 1,
            "code": "demo_code"
        }
    ],
    "kind": "full"
}



```

**Comment by tombh** [comment]
So in theory Pygls already supports this? In which case you'd like to see a more formal API that makes it easier?

**Comment by karthiknadig** [comment]
> So in theory Pygls already supports this? In which case you'd like to see a more formal API that makes it easier?

Going forward this would be my recommendation. We are also switching servers in few other languages to named pipes as well. Random packages injecting themselves using pth files and printing to stdout has been a pain to detect and debug. Turns out this is a common problem for most languages.

**Comment by tombh** [comment]
Maybe we could even default to named pipes!? I see that there's a package that can create cross-platform named pipes: https://pypi.org/project/namedpipe Whether we'd use that or not, at least demonstrates that it's possible to create named pipes without user input. Of course, we can also have an option for providing a pre-existing pipe.

**Comment by karthiknadig** [comment]
For the purposes I have mentioned we don't need to create one. Currently creating a named pipe in python for cross-platform requires calling native APIs from python. But consuming a pipe created elsewhere is easier.

You would likely only use it to test the `namedpipe` bits.

**Comment by karthiknadig** [comment]
Closing this for now, I will re-open this and provide an implementation when the client side is ready.
See https://github.com/microsoft/vscode-languageserver-node/issues/1351

Basically, on windows the LS client in VS Code uses NamedPipes, but on Unix it uses Unix Domain Socket (a socket with AF_UNIX set when creating it). My assumption when I created this was that it would be using `fifo` (named pipe in unix). We need to do some work on LS Client side before we can fully rely on `fifo`. so, for now this can be closed.
